zip-0315.rst

```diff
@@ -1,9 +1,620 @@
+value received from an untrusted note may require the user to request that
+funds are re-sent.
+
+
+Categories of TXOs according to spendability
+--------------------------------------------
+
+A TXO is *spendable*, relative to a given block chain and wallet state,
+if and only if all of the following are true in that state:
+
+* the TXO is unspent;
+* the TXO is not committed to be spent in another transaction created
+  by this wallet; and
+* the wallet has the TXO's spending key.
+
+A TXO is *confirmed-spendable*, relative to a given block chain and
+wallet state, if and only if all of the following are true in that state:
+
+* the wallet is synchronized; and
+* the TXO is spendable; and
+* either auto-shielding is disabled or the TXO is shielded; and
+* the TXO is trusted and has at least the required confirmations for
+  trusted TXOs, or it is untrusted and has at least the required
+  confirmations for untrusted TXOs.
+
+A TXO is *unconfirmed-spendable*, relative to a given block chain and
+wallet state, if and only if the TXO is spendable but is not
+confirmed-spendable in that state.
+
+A TXO is *watch-only* if and only if the wallet has its full viewing key
+(or address in the case of a transparent TXO) but not its spending key.
+
+A wallet MUST NOT attempt to spend a TXO that is not confirmed-spendable.
+
+Note: the definition of a TXO includes outputs in mempool transactions
+that are unconflicted from the perspective of the wallet.
+
+
+Reporting of balances
+---------------------
+
+Wallets SHOULD report:
+
+* Confirmed-spendable balance.
+* Pending balance, *or* total balance.
+
+These are calculated as follows:
+
+* The confirmed-spendable balance is the sum of values of
+  confirmed-spendable TXOs.
+* The pending balance is the sum of values of unconfirmed-spendable TXOs.
+* The total balance is the confirmed-spendable balance plus the pending
+  balance.
+
+Note: the definition of "confirmed-spendable" above ensures that:
+
+* if auto-shielding is enabled, transparent funds will be reported in
+  the pending or total balance, but not in the confirmed-spendable
+  balance;
+* if the wallet is not synchronized, the confirmed-spendable balance
+  will be zero.
+
+If auto-shielding is disabled, the wallet MAY report shielded and
+transparent balances separately. If it does so, it MUST make clear
+whether each reported balance corresponds to a confirmed-spendable,
+pending, or total subset of funds.
+
+Rationale for reporting of balances
+'''''''''''''''''''''''''''''''''''
+
+If auto-shielding is disabled, then separate shielded and transparent
+balances (and potentially, for expert users, separate shielded balances
+per pool) can constitute useful information. If auto-shielding is enabled
+then the wallet can and will automatically spend transparent TXOs in
+order to shield them, and so transparent TXOs need to be presented as
+pending, not as part of the balance spendable by the user.
+
```

Comment:
Also take into account auto-migration

Reply:
Taken into account by "Potentially, for expert users, separate shielded balances per pool could also be useful." I don't think we need to qualify that by whether auto-migration is enabled. I'm skeptical that it is ever useful other than for debugging, tbh.